app-backend/datamodel/src/main/scala/Scope.scala

```diff
@@ -0,0 +1,308 @@
+          Failure(
+            DecodingFailure(
+              s"Cannot build an action from split string: $result",
+              List()
+            )
+          )
+      }
+  }
+}
+
+sealed abstract class Scope {
+  val actions: Set[Action]
+}
+
+sealed class SimpleScope(
+    val actions: Set[Action]
+) extends Scope {
+  override def toString: String = actions map { _.repr } mkString (";")
+}
+
+object SimpleScope {
+  def unapply(simpleScope: SimpleScope): Option[Set[Action]] =
+    Some(simpleScope.actions)
+
+  def fromEithers(
+      actions: List[Either[DecodingFailure, Action]]
+  ): Either[DecodingFailure, SimpleScope] = {
+    val failures = actions.filter(_.isLeft)
+    if (!failures.isEmpty) {
+      Left(DecodingFailure(failures.mkString(";"), List()))
+    } else {
+      actions.sequence map { acts =>
+        new SimpleScope(acts.toSet)
+      }
+    }
+  }
+}
+
+sealed class ComplexScope(scopes: Set[Scope]) extends Scope {
+  val actions = scopes flatMap { _.actions }
+}
+
+object ComplexScope {
+  private[datamodel] def apply(scopes: Scope*): ComplexScope =
+    new ComplexScope(scopes.toSet)
+
+  def unapply(complexScope: ComplexScope): Option[Set[Action]] =
+    Some(complexScope.actions)
+}
+
+object Scope {
+
+  implicit val eqScope = new Eq[Scope] {
+    def eqv(x: Scope, y: Scope): Boolean =
+      x.actions == y.actions
+  }
+
+  implicit val monoidScope = new Monoid[Scope] {
+    def empty = Scopes.NoAccess
+    def combine(x: Scope, y: Scope): Scope =
+      ComplexScope(x, y)
+  }
+
+  implicit val encScope: Encoder[Scope] = new Encoder[Scope] {
+    def apply(thing: Scope): Json = thing match {
+      case Scopes.Uploader            => Json.fromString("uploader")
+      case Scopes.AnalysesCRUD        => Json.fromString("analyses:crud")
+      case Scopes.AnalysesMultiPlayer => Json.fromString("analyses:multiplayer")
+      case Scopes.DatasourcesCRUD     => Json.fromString("datasources:crud")
+      case Scopes.OrganizationsUserAdmin =>
+        Json.fromString("organizations:userAdmin")
+      case Scopes.ProjectExport      => Json.fromString("projects:exportFullAccess")
+      case Scopes.ProjectsFullAccess => Json.fromString("projects:fullAccess")
+      case Scopes.RasterFoundryOrganizationAdmin =>
+        Json.fromString("organizations:admin")
+      case Scopes.RasterFoundryTeamAdmin => Json.fromString("teams:admin")
+      case Scopes.ScenesCRUD             => Json.fromString("scenes:crud")
+      case Scopes.ScenesMultiPlayer      => Json.fromString("scenes:multiplayer")
+      case Scopes.ShapesFullAccess       => Json.fromString("shapes:fullAccess")
+      case Scopes.TeamsEdit              => Json.fromString("teams:edit")
+      case Scopes.TemplatesCRUD          => Json.fromString("templates:crud")
+      case Scopes.TemplatesMultiPlayer =>
+        Json.fromString("templates:multiplayer")
+      case Scopes.UploadsCRUD => Json.fromString("uploads:crud")
+      case SimpleScope(actions) =>
```

Comment:
One downside is that the possibility of ad hoc scopes means that we don't get incomplete patterns if we leave out one of the canned policies. I didn't come up with a good idea for how to make that happen, but presumably, since the canned policies get magic strings, the failure will happen in the circe codec tests and we'll catch it before it gets deployed anywhere.